GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,211 @@
+                this.pipeClient = null;
+                this.writerThread = null;
+            }
+
+            this.isDisposed = true;
+        }
+
+        protected void RaiseStateChanged(QueuedPipeStringWriterState state, string message, Exception ex)
+        {
+            this.StateChanged?.Invoke(this, new QueuedPipeStringWriterStateEventArgs(state, message, ex));
+        }
+
+        private void ThreadProc()
+        {
+            while (true)
+            {
+                // Drain the queue of all messages currently in the queue
+                while (!this.queue.IsEmpty)
+                {
+                    string message;
+                    if (this.queue.TryDequeue(out message))
+                    {
+                        this.WriteMessage(message);
+                    }
+                }
+
+                // Break from the loop to terminate the thread
+                if (!this.isRunning)
+                {
+                    break;
+                }
+
+                // There are no more messages in the queue so lets wait/sleep until we are
+                // explicitly told there are new messages, or a period of time has elapsed.
+                this.queuePulse.WaitOne(timeout: PollingInterval);
+            }
+
+            this.RaiseStateChanged(QueuedPipeStringWriterState.Stopped, null, null);
+        }
+
+        private void WriteMessage(string message)
+        {
+            // Create pipe if this is the first message, or if the last connection broke for any reason
+            if (this.pipeClient == null)
+            {
+                try
+                {
+                    // Create a new pipe stream instance using the provided factory
+                    NamedPipeClientStream pipe = this.pipeFactory();
+
+                    // Specify a instantaneous timeout because we don't want to hold up the rest of the
+                    // application if the pipe is not available; we will just drop this event.
+                    pipe.Connect(timeout: 0);
+
+                    // Keep a hold of this connected pipe for future messages
+                    this.pipeClient = pipe;
+                }
+                catch (Exception ex)
+                {
+                    this.RaiseStateChanged(QueuedPipeStringWriterState.Failing, message, ex);
+                    return;
+                }
+            }
+
+            try
+            {
+                // If we're in byte/stream transmission mode rather than message mode
+                // we should signal the end of each message with a line-feed (LF) character.
+                if (this.pipeClient.TransmissionMode == PipeTransmissionMode.Byte)
+                {
+                    message += '\n';
+                }
+
+                byte[] data = Encoding.UTF8.GetBytes(message);
+                this.pipeClient.Write(data, 0, data.Length);
+
+                this.RaiseStateChanged(QueuedPipeStringWriterState.Healthy, message, null);
```

Comment:
Yes, it does.. but in `JsonTracer` we de-dup any changes to overall `EventListener` health.

Reply:
I'll also add basic de-duping here in `QueuedPipeStringWriter`. It won't be thread safe since the only place that does a `RaiseStateChanged(..)` is the single background thread.